toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -339,7 +339,34 @@ func CreatePartitions(diskDevPath string, disk configuration.Disk, rootEncryptio
 	// Partitions assumed to be defined in sorted order
 	for idx, partition := range disk.Partitions {
 		partitionNumber := idx + 1
-		partDevPath, err := CreateSinglePartition(diskDevPath, partitionNumber, partitionTableType.String(), partition)
+		partType := "primary"
+
+		// MBR only allows up to four primary partitions, thus check here whether
+		// we need to create extended partition or not in case more than four partitions are specified
+		if partitionNumber >= 4 && len(disk.Partitions) > 4 && partitionTableType == configuration.PartitionTableTypeMbr {
+			if partitionNumber == 4 {
+				partType = "extended"
+
+				// Create a new partition object for extended partition
+				extendedPartition := configuration.Partition{}
+				extendedPartition.ID = "extended"
+				extendedPartition.Start = partition.Start
+				extendedPartition.End = disk.Partitions[len(disk.Partitions)-1].End
+
+				partDevPath, err := CreateSinglePartition(diskDevPath, partitionNumber, partitionTableType.String(), extendedPartition, partType)
+				if err != nil {
+					logger.Log.Warnf("Failed to create extended partition")
+					return partDevPathMap, partIDToFsTypeMap, encryptedRoot, readOnlyRoot, err
+				}
+				partIDToFsTypeMap[extendedPartition.ID] = ""
+				partDevPathMap[extendedPartition.ID] = partDevPath
+			}
+
+			partType = "logical"
+			partitionNumber = partitionNumber + 1
```

Comment:
Could we extract this `if` to a separate function. It's getting a bit lengthy IMO.

Reply:
When I look at this more, I'm thinking if we could do something like this (I'm not married to function names):

```
[Before the loop]
usingExtendedPartition := needsExtendedPartition(...)
if usingExtendedParittion {
    [Create extended partition]
}

for idx, partition := range disk.Partitions {
    partType, partitionNumber := inferePartitionDetails(idx, partitionNumber, usingExtendedParittion)
[Unmodified code from before this PR]
```

I think this would give us a cleaner separation of the special case for the extended partition and the more "regular" flow.

Reply:
During local testing, parted will throw errors when attempting to create partitions out of order. Needs further investigation.